DFC.ServiceTaxonomy.VersionComparison/Services/DiffBuilderService.cs

```diff
@@ -0,0 +1,133 @@
+        // We could look at creating a setting for this module to handle the parts black list but not a short piece of work
+        private readonly List<string> partsBlackList = new List<string>
+        {
+            nameof(AuditTrailPart),
+            "GraphSyncPart",
+            "ContentApprovalPart"
+        };
+
+        public DiffBuilderService(IContentDefinitionManager contentDefinitionManager, IEnumerable<IPropertyService> propertyServices)
+        {
+            _propertyServices = propertyServices;
+            _contentDefinitionManager = contentDefinitionManager;
+        }
+
+        public List<DiffItem> BuildDiffList(ContentItem baseVersion, ContentItem compareVersion)
+        {
+            if (baseVersion.ContentType != compareVersion.ContentType)
+            {
+                throw new ArgumentException(
+                    $"The base content type '{baseVersion.ContentType}' is different to the compare content type '{compareVersion.ContentType}'");
+            }
+
+            var contentType = _contentDefinitionManager.GetTypeDefinition(baseVersion.ContentType);
+            var partsLIst = contentType.Parts
+                .Select(p => p)
+                .Where(p => partsBlackList.All(bl => !bl.Equals(p.Name, StringComparison.InvariantCultureIgnoreCase)))
+                .ToList();
+            var baseJObject = JObject.FromObject(baseVersion);
+            var compareJObject = JObject.FromObject(compareVersion);
+
+            var basePartDictionary = new Dictionary<string, PropertyExtract>();
+            var comparePartDictionary = new Dictionary<string, PropertyExtract>();
+            foreach (var contentTypePartDefinition in partsLIst)
+            {
+                LoadDictionaries(basePartDictionary, contentTypePartDefinition, baseJObject);
+                LoadDictionaries(comparePartDictionary, contentTypePartDefinition, compareJObject);
+            }
+
+            return MergeDictionaries(basePartDictionary, comparePartDictionary);
+        }
+
+        private void LoadDictionaries(Dictionary<string, PropertyExtract> dictionary,
+            ContentTypePartDefinition? partDefinition, JObject? jObject)
+        {
+            var basePart = jObject?.GetValue(partDefinition?.Name) as JObject;
+            if (basePart != null)
+            {
+                LoadPropertyValues(basePart, dictionary);
+            }
+        }
+
+        private List<DiffItem> MergeDictionaries(Dictionary<string, PropertyExtract> basePartDictionary,
+            Dictionary<string, PropertyExtract> comparePartDictionary)
+        {
+            var diffList = new List<DiffItem>();
+            var keys = basePartDictionary.Keys.Union(comparePartDictionary.Keys);
+            foreach (string key in keys)
+            {
+                diffList.Add(new DiffItem{
+                    Name = basePartDictionary.ContainsKey(key) ? basePartDictionary[key].Name : comparePartDictionary[key].Name,
```

Comment:
3 lookups for the same key, should extract to a variable perhaps?

Reply:
Amended